src/types/mod.rs

```diff
@@ -144,30 +156,40 @@ impl Message {
             .unwrap()
             .as_secs();
 
-        // to compute the ttl we need to do the following
-        // - ttl = expiration - (now - msg.timestamp)
-        match now.checked_sub(self.timestamp) {
-            Some(d) => self.expiration.checked_sub(d).map(Duration::from_secs),
+        // (ts+exp) - now
+        match (self.timestamp + self.expiration).checked_sub(now) {
             None => None,
+            Some(0) => None,
+            Some(d) => Some(Duration::from_secs(d)),
         }
     }
 
     /// generic validation on the message
     pub fn valid(&self) -> Result<()> {
-        let ts = SystemTime::now()
-            .duration_since(SystemTime::UNIX_EPOCH)
-            .unwrap();
-        // ts has to be in the future of self.now
-        let du = match ts.checked_sub(Duration::from_secs(self.timestamp)) {
-            Some(du) => du,
-            None => bail!("message 'now' is in the future"),
-        };
-        if du.as_secs() > 60 {
-            bail!("message is too old ('{}' seconds)", du.as_secs());
+        if self.ttl().is_none() {
+            bail!("message has expired");
         }
 
         Ok(())
     }
+
+    /// age returns the now - message.timestamp
+    /// this will give how old the message was when
+    /// it was last stamped.
+    /// if timestamp is in the future, age will be MAX
+    pub fn age(&self) -> Duration {
+        let now = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap()
+            .as_secs();
+
+        // to compute the ttl we need to do the following
+        // - ttl = expiration - (now - msg.timestamp)
+        match now.checked_sub(self.timestamp) {
+            Some(d) => Duration::from_secs(d),
+            None => Duration::MAX,
```

Comment:
My idea was to mark the message as (unacceptably old) if the timestamp is in the future because it clearly means there is an issue with the message. Or at least this is what I was thinking at the time. It doesn't make much sense now to be honest. specially after your comment.

I am now wondering if we return 0 here that someone can abuse this by setting a timestamp in the future?